### PR TITLE
Overloaded augmented assignments

### PR DESCRIPTION
**Pull request by japaric**
Implements overload-able augmented/compound assignments, like `a += b` via the `AddAssign` trait, as specified in RFC [953](https://github.com/rust-lang/rfcs/blob/master/text/0953-op-assign.md)

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/check/op.rs [hunk 1]
```diff
@@ -167,14 +155,16 @@ fn check_overloaded_binop<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
                                     lhs_expr: &'tcx hir::Expr,
                                     lhs_ty: Ty<'tcx>,
                                     rhs_expr: &'tcx hir::Expr,
-                                    op: hir::BinOp)
+                                    op: hir::BinOp,
+                                    assign: bool)
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/op.rs]: Nit: can we newtype this bool? I think I prefer something like:

```
struct IsAssign(bool)
```

and then in the call `foo(..., IsAssign(true))`, though `IsAssign::Yes` is also ok.

File: src/librustc_typeck/check/op.rs [hunk 2]
```diff
@@ -231,27 +228,51 @@ pub fn check_user_unop<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
     }
 }
 
-fn name_and_trait_def_id(fcx: &FnCtxt, op: hir::BinOp) -> (&'static str, Option<DefId>) {
+fn name_and_trait_def_id(fcx: &FnCtxt,
+                         op: hir::BinOp,
+                         assign: bool)
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/op.rs]: Nit: newtype here too.

File: src/librustc_typeck/check/writeback.rs [hunk 3]
```diff
@@ -90,24 +90,51 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
...
-
-                // weird but true: the by-ref binops put an
-                // adjustment on the lhs but not the rhs; the
-                // adjustment for rhs is kind of baked into the
-                // system.
-                if !hir_util::is_by_value_binop(op.node) {
-                    self.fcx.inh.tables.borrow_mut().adjustments.remove(&lhs.id);
+        match e.node {
+            hir::ExprBinary(ref op, ref lhs, ref rhs) |
+                hir::ExprAssignOp(ref op, ref lhs, ref rhs) => {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/writeback.rs]: nit: indentation looks off

File: src/librustc_typeck/check/writeback.rs [hunk 4]
```diff
@@ -90,24 +90,51 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
...
+                                self.fcx.inh.tables.borrow_mut().adjustments.remove(&lhs.id);
+                            },
+                            _ => {},
+                        }
+                    } else {
+                        let tcx = self.tcx();
+
+                        if let hir::ExprAssignOp(..) = e.node {
+                            if !tcx.sess.features.borrow().augmented_assignments &&
+                                !self.fcx.expr_ty(e).references_error() {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/writeback.rs]: Nit: the body of the if and the condition ought not be aligned the same without an intervening `{`, imo. :) Can you move the opening brace to the next line, or else modify the indent, or something? I personally often do this but I think I am the only one that does so: ;)

```rust
if
    cond1 &&
    cond2
{
    ...
}
```

**Comment by nikomatsakis** [comment]
This looks good. r+ modulo nits -- can you also add a test that you can't implement `AddAssign` and friends without the suitable feature gate? (Another related question -- do we really want two distinct feature gates? Seems like the traits and associated syntax live or die together?)

Also: I think it may be worth waiting to land this until the beta is released, just in case it introduces accidental breakage.

**Comment by japaric** [comment]
@nikomatsakis I've addressed all your comments (I think)

> do we really want two distinct feature gates? Seems like the traits and associated syntax live or die together?

I did try using the same feature name for the language feature and for the traits, but it didn't work. What I observed is that when I added `#![feature(augmented_assignment)]` to some source file, the language feature was enabled, i.e. `a += b` didn't raised an error; but implementing the `OpAssign` trait still raised the "unstable library feature 'augmented_assignments'" error. Which seems to indicate there's a bug when a language feature and a library feature have the same name: `#![feature(foo)]` enables the language feature `foo`, but not the library feature `foo` when it should enable both.

> Also: I think it may be worth waiting to land this until the beta is released, just in case it introduces accidental breakage.

Sounds good to me

**Comment by nikomatsakis** [comment]
@japaric ok feel free to r=me once rebased

**Pull request merged**